### PR TITLE
do not show prefix and suffix for minor log axes

### PR DESCRIPTION
**Pull request by archmoj**
Hopefully fixes https://github.com/plotly/plotly.js/issues/3250 & friends (i.e. https://github.com/plotly/plotly.js/issues/3689).

@plotly/plotly_js

### CODE REVIEW

File: src/plots/cartesian/axes.js [hunk 1]
```diff
@@ -571,34 +571,44 @@ axes.calcTicks = function calcTicks(ax) {
...
             x = axes.tickIncrement(x, ax.dtick, axrev, ax.calendar)) {
         // prevent infinite loops - no more than one tick per pixel,
         // and make sure each value is different from the previous
-        if(vals.length > maxTicks || x === xPrevious) break;
+        if(tickVals.length > maxTicks || x === xPrevious) break;
         xPrevious = x;
 
-        vals.push(x);
+        var minor = false;
+        if(isLog && (x !== (x | 0))) {
```
Comment (etpinard) [inline src/plots/cartesian/axes.js]: What does `x !== (x | 0)` mean?

Comment (etpinard) [reply src/plots/cartesian/axes.js]: Ha I see, it means _if `x` isn't an integer_, right?

Comment (etpinard) [reply src/plots/cartesian/axes.js]: In https://github.com/plotly/plotly.js/issues/3250 , @alexcjohnson wrote

> seems like for dtick='D1' and 'D2', tickprefix (and ticksuffix if given) should only apply to the "complete" labels at the powers of 10, not the single-digit labels between.

@archmoj can you confirm that _should only to "complete" labels at the powers of 10" and _if `x` isn't an integer_ are equivalent here?

**Comment by etpinard** [comment]
Thanks very much for taking this on @archmoj !

I'm curious, you wrote that this PR fixes #3250 & friends, which _friend issues_ are you referring to exactly?

**Comment by archmoj** [comment]
> Thanks very much for taking this on @archmoj !
> 
> I'm curious, you wrote that this PR fixes #3250 & friends, which _friend issues_ are you referring to exactly?

@etpinard 
This one: https://github.com/plotly/plotly.js/issues/3689.

File: src/plots/cartesian/axes.js (hunk 1 above)
Comment (archmoj) [reply src/plots/cartesian/axes.js]: > Ha I see, it means _if `x` isn't an integer_, right?

Correct.

Comment (archmoj) [reply src/plots/cartesian/axes.js]: > In #3250 , @alexcjohnson wrote
> 
> > seems like for dtick='D1' and 'D2', tickprefix (and ticksuffix if given) should only apply to the "complete" labels at the powers of 10, not the single-digit labels between.
> 
> @archmoj can you confirm that _should only to "complete" labels at the powers of 10" and _if `x` isn't an integer_ are equivalent here?

Yes. It seems so.

Comment (etpinard) [reply src/plots/cartesian/axes.js]: Have you tried your branch together with `dtick` of the `L<n>` and `D<n>` families?

https://github.com/plotly/plotly.js/blob/56dd829fbaa2ba5e254bc6870b380ef333bf30e5/src/plots/cartesian/layout_attributes.js#L296-L300

Comment (archmoj) [reply src/plots/cartesian/axes.js]: > Have you tried your branch together with `dtick` of the `L<n>` and `D<n>` families?
> 
> https://github.com/plotly/plotly.js/blob/56dd829fbaa2ba5e254bc6870b380ef333bf30e5/src/plots/cartesian/layout_attributes.js#L296-L300

@etpinard Good call! 
Fixed new logic & improved the test in https://github.com/plotly/plotly.js/pull/4064/commits/27067a6c8f45dc0fbfa19dbaad0393047773601c.

File: src/plots/cartesian/axes.js [hunk 2]
```diff
@@ -571,34 +571,44 @@ axes.calcTicks = function calcTicks(ax) {
...
 
     // return the full set of tick vals
-    var vals = [];
+    var tickVals = [];
     if(ax.type === 'category' || ax.type === 'multicategory') {
         endTick = (axrev) ? Math.max(-0.5, endTick) :
             Math.min(ax._categories.length - 0.5, endTick);
     }
 
+    var isDLog = (ax.type === 'log') && !(isNumeric(ax.dtick) || ax.dtick.charAt(0) === 'L');
```
Comment (etpinard) [inline src/plots/cartesian/axes.js]: @archmoj Nice!

Could you share a screenshot of what your `log-axis_no-minor_suffix-prefix` looked like before 27067a6c8f45dc0fbfa19dbaad0393047773601c 

Thanks!

File: test/image/mocks/log-axis_no-minor_suffix-prefix.json [hunk 3]
```diff
@@ -0,0 +1,117 @@
+{
+  "data": [
+    {
+      "y": [
```
Comment (etpinard) [inline test/image/mocks/log-axis_no-minor_suffix-prefix.json]: Could you set the trace `name` to their corresponding `yaxis?.dtick` value to improve the mock?

File: src/plots/cartesian/axes.js (hunk 2 above)
Comment (archmoj) [reply src/plots/cartesian/axes.js]: before4064

Comment (etpinard) [reply src/plots/cartesian/axes.js]: Awesome - thanks!!

File: test/image/mocks/log-axis_no-minor_suffix-prefix.json (hunk 3 above)
Comment (archmoj) [reply test/image/mocks/log-axis_no-minor_suffix-prefix.json]: Good call. Done in https://github.com/plotly/plotly.js/pull/4064/commits/cb521f5d29e539527075a54c1900c5b232a19791.

**Comment by etpinard** [comment]
This fix is looking good to me :dancer: 

Nice work @archmoj !

**Pull request merged**